lib/parser/models.ex

```diff
@@ -0,0 +1,385 @@
+                           nil ->
+                             nil
+                           default ->
+                             cast(type, default)
+                         end
+        %Field{id: id,
+               type: type,
+               name: atomify(name),
+               required: required,
+               default: casted_default}
+      end
+
+      def build_field_list(parent_name, fields) do
+        fields
+        |> update_ids(parent_name)
+        |> validate_ids(parent_name)
+      end
+
+      defp validate_ids(fields, name) do
+        dupes = fields
+        |> Enum.group_by(&(&1.id))
+        |> Enum.filter(fn {_, v} -> length(v) > 1 end)
+
+        unless Enum.empty?(dupes) do
+          {id, fields} = List.first(dupes)
+
+          names = fields
+          |> Enum.map(&("#{name}.#{&1.name}"))
+          |> Enum.join(", ")
+
+          raise "Error: #{names} share field number #{id}."
+        end
+
+        fields
+      end
+
+      defp update_ids(fields, parent_name) do
+        alias Mix.Shell.IO
+        fields
+        |> Enum.with_index
+        |> Enum.map(fn
+          {field=%__MODULE__{}, idx} ->
+            case field.id do
+              nil ->
+                IO.info "Warning: id not set for field '#{parent_name}.#{field.name}'."
+                %__MODULE__{field | id: idx + 1}
+              _ ->
+                field
+            end
+        end)
+      end
+
+    end
+
+
+    defmodule Exception do
+      @moduledoc """
+      A thrift exception
+
+      """
+      defstruct fields: %{}, name: nil
+      import Thrift.Parser.Conversions
+      alias Thrift.Parser.Models.Field
+
+      def new(name, fields) do
+        ex_name = atomify(name)
+        updated_fields = Field.build_field_list(ex_name, fields)
+
+        %Exception{name: ex_name, fields: updated_fields}
+      end
+    end
+
+    defmodule Struct do
+      @moduledoc """
+      A thrift struct
+
+      The basic datastructure in Thrift, structs have aa name and a field list.
+      """
+      defstruct name: nil, fields: %{}
+      import Thrift.Parser.Conversions
+      alias Thrift.Parser.Models.Field
+
+      def new(name, fields) do
+        struct_name = atomify(name)
+        fields = Field.build_field_list(struct_name, fields)
+
+        %Struct{name: struct_name, fields: fields}
+      end
+    end
+
+    defmodule Union do
+      @moduledoc """
+      A thrift union
+
+      Unions can have one field set at a time.
+      """
+      defstruct name: nil, fields: %{}
+      import Thrift.Parser.Conversions
+      alias Thrift.Parser.Models.Field
+
+      def new(name, fields) do
+        name = atomify(name)
+        fields = Field.build_field_list(name, fields)
+        |> Enum.map(fn(field=%Field{}) ->
+          # According to thrift docs, unions have implicitly optional
+          # fields. See https://thrift.apache.org/docs/idl#union
+          %Field{field | required: false}
```

Comment:
It can't, I get an error about this not supporting the Access protocol. Map.put works, but I think this is the preferred method for updating structs.